### DISCUSSION TITLE
Custom sound file grayed out

### DISCUSSION THREAD

**Discussion by Navid200** [NightscoutFoundation/xDrip#1983] [General]
This discussion explains how to recreate a problem.
I have an open PR: https://github.com/NightscoutFoundation/xDrip/pull/1827 to address the problem.
I didn't want to clutter the PR.  So, I am using this discussion for that purpose.

**Comment by Navid200** [comment]
This is my Android Studio:
![Untitled](https://user-images.githubusercontent.com/51497406/152706228-6b61faec-373e-4052-8d16-ec89ab5ce747.png)

  
I added this virtual device:
![Screenshot 2022-02-06 183323](https://user-images.githubusercontent.com/51497406/152706333-cca92ba0-721b-42f4-8649-e0ab3e044abb.png)  
  
  
Brought up the virtual device and downloaded xDrip latest Nightly:
![Screenshot 2022-02-06 185647](https://user-images.githubusercontent.com/51497406/152707090-c37ec9db-fb04-4599-952e-b4eb6c2c0d9a.png)

Installed xDrip in the Pixel3 virtual device:
![Screenshot 2022-02-06 185926](https://user-images.githubusercontent.com/51497406/152707203-bc61921c-4fed-4752-8ec1-82b161177a2a.png)

Downloaded an mp3 file:
![Screenshot 2022-02-06 191005](https://user-images.githubusercontent.com/51497406/152707617-56ab5b08-56f3-45a0-a392-b4e5612d5255.png)

The sample mp3 file (file_example_MP3_700KB.mp3) was downloaded to the Download folder: 
![Screenshot 2022-02-06 191244](https://user-images.githubusercontent.com/51497406/152707730-2c1b6689-a375-42ac-8614-c06d4dd4b3ca.png)

Use the file browser and go to the Download folder and tap on the downloaded file to see the title of the music:
![Screenshot 2022-02-06 195730](https://user-images.githubusercontent.com/51497406/152709652-87c3566c-0e6d-40f0-9190-cbb5ae066e9f.png)


Select xDrip Settings:
![Screenshot 2022-02-06 193118](https://user-images.githubusercontent.com/51497406/152708512-8610ce5c-3583-4ff6-8952-d65db1cd0e7c.png)

Select Alarms and Alerts:
![Screenshot 2022-02-06 193400](https://user-images.githubusercontent.com/51497406/152708586-46afb9be-4ba0-45f3-ad25-271cceef200e.png)
  
Select Glucose Level Alerts List:
![Screenshot 2022-02-06 193517](https://user-images.githubusercontent.com/51497406/152708655-deaef0a5-eb36-4388-a289-4ddc6caf8b4b.png)
  
  
Long press one of the existing alerts:
![Screenshot 2022-02-06 193654](https://user-images.githubusercontent.com/51497406/152708708-729f8af9-d920-4c79-ac3c-fae95e85b3d3.png)
  
  
Choose File:
![Screenshot 2022-02-06 193827](https://user-images.githubusercontent.com/51497406/152708777-8f99a2bf-216e-4770-a213-34ec82cceefd.png)
  
  
Choose Custom Sound/Alarm:  
![Screenshot 2022-02-06 193922](https://user-images.githubusercontent.com/51497406/152708859-2559fbaf-e323-4d62-b5ee-9fc87cd2ede3.png)
  
OK allow permission:
![Screenshot 2022-02-06 194102](https://user-images.githubusercontent.com/51497406/152708899-f5150fa6-85f5-40c4-bcca-5443cae611f6.png)
  
Allow:
![Screenshot 2022-02-06 194159](https://user-images.githubusercontent.com/51497406/152708932-23eeae2e-6aa0-4e4e-aba8-81c26fd9d177.png)

Go back to xDrip and tap on Choose file again and choose Custom Sound/Alarm again.
![Screenshot 2022-02-06 194425](https://user-images.githubusercontent.com/51497406/152709095-b2edbecf-c052-4e56-bd4c-dff31fd2d95e.png)

Use the hamburger menu button and go to the internal drive top level:
![Screenshot 2022-02-06 194902](https://user-images.githubusercontent.com/51497406/152709226-55783f09-f2f1-419d-bdf0-d1193512c912.png)


Go to Download:
![Screenshot 2022-02-06 195004](https://user-images.githubusercontent.com/51497406/152709272-6f5a6136-ef82-4df2-b9a6-ef1996046351.png)
The mp3 file is grayed out and cannot be selected.

Now, copy the file to the Notification folder at the internal storage top level.
  
Go back to the alert edit page and tap on choose file again.  This time, choose System Sound/Alarm.
![Screenshot 2022-02-06 200127](https://user-images.githubusercontent.com/51497406/152709789-d1d39b7e-10c7-46ff-b59d-ce671621fbba.png)

Choose "My Sounds":
![Screenshot 2022-02-06 200305](https://user-images.githubusercontent.com/51497406/152709891-0f18ef65-fda1-4f9a-8230-9e41ca1c3b3a.png)


You can see the file.  This is the work-around I have been using and have been telling everyone who faces the same problem to use.
I can continue to do that.  The problem is that everyone who experiences the problem may not post and ask about it.  They may just give up and use a different app.
We can use custom files.  We just need to place them in the right folder and they will be presented to us under the system sound/alarm option.

**Comment by tzachi-dar** [comment]
Thanks for your wonderful explanation.
I was able to verify what you have seen. It seems that there are a few things that we can do to solve this (behind what you already suggest to people):
1) Install a file manager on the phone, and use it for choosing. For example: file manger plus.
2) Currently we set the file type to fileIntent.setType("audio/mpeg3"); changing this to fileIntent.setType("*/*"); will allow the user to choose all files (we will need to filter the files later). This is explained with more details here: https://stackoverflow.com/questions/27723975/files-are-unselectable-after-calling-action-get-content-intent
3) We can implment our own file choosing dialog.

I'll try to look at the second option tonight to see if it works well.

**Reply by tolot27** [reply]
The second options looks like the solution we should implement.

**Comment by tzachi-dar** [comment]
Ok, based on https://stackoverflow.com/questions/13209494/how-to-get-the-full-file-path-from-uri/60642994#answer-60642994 it seems that there is a working implmentation of #2.
I'll check it on more devices and hopefully this will work.

**Comment by tzachi-dar** [comment]
I have created a PR that shold fix the issue. https://github.com/NightscoutFoundation/xDrip/pull/1986

@Navid200 will be intersting if you can check and see that it is working well.

**Reply by Navid200** [reply]
I wanted to test that PR.  But, I realized that even if it worked, it will only work for the Glucose alert list and not for Persistent High alert or the Forecast Low alert.

The solution that works consistently for all alerts is to simply place the mp3 file in the Notifications folder.
That is a very simple solution.  The only thing we need to is to remove the option that doesn't consistently work for everyone.
